### PR TITLE
feat: set allocations

### PR DESCRIPTION
**Pull request by 8sunyuan**
Draft for changing allocate/deallocate to just `setAllocations` which can specify absolute magnitude values vs previously we had `magnitudeDiffs`. Depending on the current magnitude value assigned for the (op, Strategy, opSet) tuple, it will either allocate/deallocate accordingly.

NOTE: in `setAllocations` any magnitude values that are equal to the current value are ignored as it is neither a allocation or deallocation. 

Additional changes included:
- expectedTotalMagnitude check
- initializeable allocationDelays that could be set one time by an operator's allocator
- Checkpoints library optimization: reuse length check in `upperLookupRecentWithPos`

### CODE REVIEW

File: src/contracts/core/AVSDirectory.sol [hunk 1]
```diff
@@ -27,7 +27,12 @@ contract AVSDirectory is
     /// @dev BIPS factor for slashable bips
     uint256 internal constant BIPS_FACTOR = 10_000;
     /// @dev Delay before allocations take effect and how long until deallocations are completable
-    uint32 public constant ALLOCATION_DELAY = 21 days;
+    uint32 public constant DEFAULT_ALLOCATION_DELAY = 17.5 days;
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:30]: 21 prob better here

File: src/contracts/core/AVSDirectory.sol [hunk 2]
```diff
@@ -282,7 +287,7 @@ contract AVSDirectory is
             });
         }
 
-        uint32 effectTimestamp = uint32(block.timestamp) + ALLOCATION_DELAY;
+        uint32 effectTimestamp = uint32(block.timestamp) + DEALLOCATION_DELAY;
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:290]: needs to be custom allocation delay?

File: src/contracts/core/AVSDirectory.sol [hunk 3]
```diff
@@ -560,14 +709,13 @@ contract AVSDirectory is
...
-            (uint224 value, uint256 pos) = _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i]
+            (uint224 value, uint256 pos, uint256 length) = _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i]
                 .operatorSetId].upperLookupRecentWithPos(uint32(block.timestamp));
             // no checkpoint exists if value == 0 && pos == 0, so check the negation before checking if there is a
             // a pending allocation
             if (value != 0 || pos != 0) {
                 require(
-                    pos + MAX_PENDING_UPDATES
-                        == _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId].length(),
+                    pos + MAX_PENDING_UPDATES == length,
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:718]: <=?

File: src/contracts/core/AVSDirectory.sol [hunk 4]
```diff
@@ -344,6 +349,128 @@ contract AVSDirectory is
...
+        OperatorSet[] operatorSets;
+        uint64[] magnitude;
+    }
+
+    function setAllocations(
+        address operator,
+        MagnitudeAllocation[] calldata allocations,
+        SignatureWithSaltAndExpiry calldata allocatorSignature
+    ) external {
+        // completable timestamp for deallocations
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:364]: verify msg.sender

File: src/contracts/core/AVSDirectory.sol [hunk 5]
```diff
@@ -448,6 +575,28 @@ contract AVSDirectory is
...
+    function initializeAllocationDelay(
+        address operator,
+        uint32 delay
+    ) external {
+        require(
+            delegation.isOperator(operator),
+            "AVSDirectory.initializeAllocationDelay: operator not registered to EigenLayer yet"
+        );
+        require(
+            delegation.allocator(operator) == msg.sender,
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:587]: doesn't exist, no?

File: src/contracts/core/AVSDirectory.sol [hunk 6]
```diff
@@ -344,6 +349,128 @@ contract AVSDirectory is
...
+                // Newly configured magnitude is greater than current value. 
+                // Therefore we handle this as an allocation
+
+                // 1. ensure only 1 pending allocation at a time
+                // read number of checkpoints after current timestamp
+                // no checkpoint exists if value == 0 && pos == 0, so check the negation before checking if there is a
+                // a pending allocation
+                if (currentMagnitude != 0 || pos != 0) {
+                    require(
+                        pos + MAX_PENDING_UPDATES == length,
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:450]: <=

File: src/contracts/core/AVSDirectory.sol [hunk 7]
```diff
@@ -344,6 +349,128 @@ contract AVSDirectory is
...
+                        "AVSDirectory.queueAllocations: exceed max pending allocations allowed for op, opSet, strategy"
+                    );
+                }
+                // 2. allocate magnitude which will take effect in the future 21 days from now
+                _magnitudeUpdate[operator][allocation.strategy][opSets[i].avs][opSets[i].operatorSetId].push({
+                    key: allocationEffectTimestamp,
+                    value: allocation.magnitude[i]
+                });
+                // 3. decrement free magnitude by incremented amount
+                // TODO??? could let this underflow instead of explicit check
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:460]: explicit check is good

File: src/contracts/core/AVSDirectory.sol [hunk 8]
```diff
@@ -344,6 +349,128 @@ contract AVSDirectory is
...
+        }
+    }
+
+    function _setAllocations(
+        address operator,
+        MagnitudeAllocation calldata allocation,
+        uint32 allocationEffectTimestamp,
+        uint32 deallocationCompletableTimestamp
+    ) internal {
+        uint64 currentFreeMagnitude = freeMagnitude[operator][allocation.strategy];
```
Comment (gpsanant) [inline src/contracts/core/AVSDirectory.sol:400]: newFreeMagnitude

File: src/contracts/core/AVSDirectory.sol (hunk 3 above)
Comment (8sunyuan) [reply src/contracts/core/AVSDirectory.sol:718]: Actually think it should be >=

File: src/contracts/core/AVSDirectory.sol (hunk 8 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

File: src/contracts/core/AVSDirectory.sol (hunk 7 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

File: src/contracts/core/AVSDirectory.sol (hunk 6 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

File: src/contracts/core/AVSDirectory.sol (hunk 5 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: not sure if we want to nuke from interface as well?

File: src/contracts/core/AVSDirectory.sol (hunk 3 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

File: src/contracts/core/AVSDirectory.sol (hunk 2 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

File: src/contracts/core/AVSDirectory.sol (hunk 1 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

File: src/contracts/core/AVSDirectory.sol (hunk 4 above)
Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol:279]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/db66da58945da71f7f53940f6dbf656a7d25d61e

Comment (gpsanant) [reply src/contracts/core/AVSDirectory.sol:279]: still need to handle signature

File: src/contracts/core/AVSDirectory.sol (hunk 5 above)
Comment (8sunyuan) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/93ba8afce48fad4f4b2d0d357dc31e190482ce88

File: src/contracts/core/AVSDirectory.sol (hunk 4 above)
Comment (8sunyuan) [reply src/contracts/core/AVSDirectory.sol:284]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/93ba8afce48fad4f4b2d0d357dc31e190482ce88
Added signature handling

File: src/contracts/core/AVSDirectory.sol (hunk 3 above)
Comment (8sunyuan) [reply src/contracts/core/AVSDirectory.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/691/commits/decb79d6b4a33c3840a23460e55d8799b6fd455e
Fixed

**Pull request merged**